DatadogRUM/Sources/Instrumentation/AppHangs/AppHangsMonitor.swift

```diff
@@ -0,0 +1,99 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+import DatadogInternal
+
+internal final class AppHangsMonitor {
+    enum Constants {
+        /// The standardized `error.message` for RUM errors describing an app hang.
+        static let appHangErrorMessage = "App Hang"
+        /// The standardized `error.type` for RUM errors describing an app hang.
+        static let appHangErrorType = "AppHang"
+        /// The standardized `error.stack` when backtrace generation was not available.
+        static let appHangStackNotAvailableErrorMessage = "Stack trace was not generated because `DatadogCrashReporting` had not been enabled."
+        /// The standardized `error.stack` when backtrace generation failed due to an internal error.
+        static let appHangStackGenerationFailedErrorMessage = "Failed to generate stack trace. This is a known issue and we work on it."
+    }
+
+    /// Watchdog thread that monitors the main queue for App Hangs.
+    private let watchdogThread: AppHangsObservingThread
+    /// Handles non-fatal App Hangs.
+    internal let nonFatalHangsHandler: NonFatalAppHangsHandler
+    /// Handles non-fatal App Hangs.
+    internal let fatalHangsHandler: FatalAppHangsHandler
```

Comment:
💡 The previous non-fatal App Hang logic went to `NonFatalAppHangsHandler`. Fatals are implemented in separate handler. This is to separate concerns as both will evolve significantly different ways (with fatal handler reaching significantly higher complexity).